br/pkg/membuf/buffer.go

```diff
@@ -172,17 +175,51 @@ func (p *Pool) NewBuffer(opts ...BufferOption) *Buffer {
 	return b
 }
 
-// Reset resets the buffer, the memory is still retained in this buffer.
+// smallObjOverheadBatch is the batch size to acquire memory from limiter. 256KB
+// can store 256KB/24B = 10922 []byte objects, or 256KB/12B = 21845 SliceLocation
+// objects.
+const smallObjOverheadBatch = 256 * 1024
+
+// recordSmallObjOverhead records the memory cost of []byte or SliceLocation into
+// pool's limiter. The caller will ensure the pool's limiter is not nil.
+func (b *Buffer) recordSmallObjOverhead(n int) {
+	if n > b.smallObjOverheadCache {
+		b.pool.limiter.Acquire(smallObjOverheadBatch)
```

Comment:
I think for each goroutine, `p.limiter.Acquire()` should be invoked only once? Otherwise, there is a risk of deadlock.

```
Total memory: 10

goroutine 1:
func AllocBytesWithSliceLocation() {
    b.allocBytesWithSliceLocation() // limiter.Acquire(9), ok
->  b.recordSmallObjOverhead()      // limiter.Acquire(2), but 9 + 2 > 10, block
}

goroutine 2:
func AllocBytesWithSliceLocation() {
->  b.allocBytesWithSliceLocation() // limiter.Acquire(9), but 9 + 9 > 10, block 
    b.recordSmallObjOverhead()
}
```

Reply:
In this lock acquisition pattern we only rely the caller don't have too many concurrency and don't acquire too many memory to avoid deadlock 😂 

We don't know how many []byte will be allocated from the `Buffer` so it can't call `Acquire` once.